# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_sentences = source_text.split('\n')
    sentence_words = [line.split() for line in source_sentences]               
    source_id_text = [[source_vocab_to_int[word] for word in line] for line in sentence_words]
        
    target_sentences = target_text.split('\n')
    sentence_words = [line.split() for line in target_sentences]
    target_id_text = [[target_vocab_to_int[word] for word in line] + [target_vocab_to_int['<EOS>']] for line in sentence_words]
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, [None, None], name = 'input')
    targets = tf.placeholder(tf.int32, [None, None], name = 'targets')
    learning_rate = tf.placeholder(tf.float32, name = 'learning_rate')
    keep_prob = tf.placeholder(tf.float32, name = 'keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, (None,), name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, (None,), name='source_sequence_length')
        
    return (inputs, targets, learning_rate, keep_prob,
            target_sequence_length, max_target_sequence_length, source_sequence_length)
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size)
        enc_cell = tf.contrib.rnn.DropoutWrapper(enc_cell, output_keep_prob=keep_prob)
        return enc_cell
    
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)

    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input,
                                              sequence_length=source_sequence_length, dtype=tf.float32)
        
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    training_helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, training_helper, encoder_state, output_layer) 
        
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder, impute_finished=True,
                                                                   maximum_iterations=max_summary_length)[0]
    
    return training_decoder_output
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)

    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, inference_helper, encoder_state, output_layer)
        
    inference_decoder_output = tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                               maximum_iterations=max_target_sequence_length)[0]
    
    return inference_decoder_output

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function    
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size)
        dec_cell = tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)
        return dec_cell
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    output_layer = Dense(target_vocab_size, kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("decode") as scope:
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_target_sequence_length, output_layer, keep_prob)
        scope.reuse_variables()
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, 
            target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], max_target_sequence_length,
            target_vocab_size, output_layer, batch_size, keep_prob)
    
    return training_decoder_output, inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    _, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, enc_embedding_size)
    
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, enc_state, target_sequence_length,
                                                                       max_target_sentence_length, rnn_size, num_layers,
                                                                       target_vocab_to_int, target_vocab_size, batch_size,
                                                                       keep_prob, dec_embedding_size)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [26]:
# Number of Epochs
epochs = 10
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 512
# Number of Layers
num_layers = 1
# Embedding Size
encoding_embedding_size = 128
decoding_embedding_size = 128
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.5
display_step = True

### Build the Graph
Build the graph using the neural network you implemented.

In [27]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [28]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [29]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    1/538 - Train Accuracy: 0.2314, Validation Accuracy: 0.3157, Loss: 4.9356
Epoch   0 Batch    2/538 - Train Accuracy: 0.2516, Validation Accuracy: 0.3157, Loss: 4.6256
Epoch   0 Batch    3/538 - Train Accuracy: 0.2604, Validation Accuracy: 0.3448, Loss: 4.3473
Epoch   0 Batch    4/538 - Train Accuracy: 0.2662, Validation Accuracy: 0.3423, Loss: 4.1641
Epoch   0 Batch    5/538 - Train Accuracy: 0.2943, Validation Accuracy: 0.3443, Loss: 4.0149
Epoch   0 Batch    6/538 - Train Accuracy: 0.3045, Validation Accuracy: 0.3475, Loss: 3.7862
Epoch   0 Batch    7/538 - Train Accuracy: 0.2879, Validation Accuracy: 0.3549, Loss: 3.6895
Epoch   0 Batch    8/538 - Train Accuracy: 0.2953, Validation Accuracy: 0.3647, Loss: 3.6301
Epoch   0 Batch    9/538 - Train Accuracy: 0.3094, Validation Accuracy: 0.3723, Loss: 3.4983
Epoch   0 Batch   10/538 - Train Accuracy: 0.2961, Validation Accuracy: 0.3773, Loss: 3.5046
Epoch   0 Batch   11/538 - Train Accuracy: 0.3234, Validation Accuracy

Epoch   0 Batch   91/538 - Train Accuracy: 0.4615, Validation Accuracy: 0.5153, Loss: 1.3048
Epoch   0 Batch   92/538 - Train Accuracy: 0.4592, Validation Accuracy: 0.5204, Loss: 1.2870
Epoch   0 Batch   93/538 - Train Accuracy: 0.4635, Validation Accuracy: 0.5117, Loss: 1.2658
Epoch   0 Batch   94/538 - Train Accuracy: 0.4697, Validation Accuracy: 0.5169, Loss: 1.2885
Epoch   0 Batch   95/538 - Train Accuracy: 0.5078, Validation Accuracy: 0.5096, Loss: 1.1541
Epoch   0 Batch   96/538 - Train Accuracy: 0.4939, Validation Accuracy: 0.5192, Loss: 1.1693
Epoch   0 Batch   97/538 - Train Accuracy: 0.4512, Validation Accuracy: 0.5256, Loss: 1.2379
Epoch   0 Batch   98/538 - Train Accuracy: 0.4901, Validation Accuracy: 0.5181, Loss: 1.1609
Epoch   0 Batch   99/538 - Train Accuracy: 0.4430, Validation Accuracy: 0.5030, Loss: 1.2379
Epoch   0 Batch  100/538 - Train Accuracy: 0.4732, Validation Accuracy: 0.4970, Loss: 1.1894
Epoch   0 Batch  101/538 - Train Accuracy: 0.4836, Validation Accuracy

Epoch   0 Batch  181/538 - Train Accuracy: 0.5326, Validation Accuracy: 0.5822, Loss: 0.8469
Epoch   0 Batch  182/538 - Train Accuracy: 0.5500, Validation Accuracy: 0.5888, Loss: 0.8171
Epoch   0 Batch  183/538 - Train Accuracy: 0.6023, Validation Accuracy: 0.5836, Loss: 0.7464
Epoch   0 Batch  184/538 - Train Accuracy: 0.5778, Validation Accuracy: 0.5820, Loss: 0.7657
Epoch   0 Batch  185/538 - Train Accuracy: 0.5633, Validation Accuracy: 0.5792, Loss: 0.7908
Epoch   0 Batch  186/538 - Train Accuracy: 0.5614, Validation Accuracy: 0.5817, Loss: 0.7834
Epoch   0 Batch  187/538 - Train Accuracy: 0.5799, Validation Accuracy: 0.5813, Loss: 0.7603
Epoch   0 Batch  188/538 - Train Accuracy: 0.5375, Validation Accuracy: 0.5790, Loss: 0.7873
Epoch   0 Batch  189/538 - Train Accuracy: 0.5705, Validation Accuracy: 0.5746, Loss: 0.7905
Epoch   0 Batch  190/538 - Train Accuracy: 0.5493, Validation Accuracy: 0.5785, Loss: 0.7981
Epoch   0 Batch  191/538 - Train Accuracy: 0.5882, Validation Accuracy

Epoch   0 Batch  271/538 - Train Accuracy: 0.5855, Validation Accuracy: 0.6168, Loss: 0.6831
Epoch   0 Batch  272/538 - Train Accuracy: 0.5732, Validation Accuracy: 0.6191, Loss: 0.7139
Epoch   0 Batch  273/538 - Train Accuracy: 0.6096, Validation Accuracy: 0.6222, Loss: 0.6796
Epoch   0 Batch  274/538 - Train Accuracy: 0.5660, Validation Accuracy: 0.6142, Loss: 0.7223
Epoch   0 Batch  275/538 - Train Accuracy: 0.5992, Validation Accuracy: 0.6246, Loss: 0.6980
Epoch   0 Batch  276/538 - Train Accuracy: 0.6062, Validation Accuracy: 0.6284, Loss: 0.6696
Epoch   0 Batch  277/538 - Train Accuracy: 0.5957, Validation Accuracy: 0.6239, Loss: 0.6686
Epoch   0 Batch  278/538 - Train Accuracy: 0.5998, Validation Accuracy: 0.6177, Loss: 0.6740
Epoch   0 Batch  279/538 - Train Accuracy: 0.6055, Validation Accuracy: 0.6310, Loss: 0.6476
Epoch   0 Batch  280/538 - Train Accuracy: 0.6408, Validation Accuracy: 0.6323, Loss: 0.6245
Epoch   0 Batch  281/538 - Train Accuracy: 0.6111, Validation Accuracy

Epoch   0 Batch  361/538 - Train Accuracy: 0.6224, Validation Accuracy: 0.6467, Loss: 0.5639
Epoch   0 Batch  362/538 - Train Accuracy: 0.6490, Validation Accuracy: 0.6461, Loss: 0.5507
Epoch   0 Batch  363/538 - Train Accuracy: 0.6378, Validation Accuracy: 0.6465, Loss: 0.5472
Epoch   0 Batch  364/538 - Train Accuracy: 0.6074, Validation Accuracy: 0.6452, Loss: 0.6028
Epoch   0 Batch  365/538 - Train Accuracy: 0.6231, Validation Accuracy: 0.6410, Loss: 0.5577
Epoch   0 Batch  366/538 - Train Accuracy: 0.6334, Validation Accuracy: 0.6506, Loss: 0.5802
Epoch   0 Batch  367/538 - Train Accuracy: 0.6475, Validation Accuracy: 0.6403, Loss: 0.5486
Epoch   0 Batch  368/538 - Train Accuracy: 0.6701, Validation Accuracy: 0.6374, Loss: 0.5166
Epoch   0 Batch  369/538 - Train Accuracy: 0.6391, Validation Accuracy: 0.6412, Loss: 0.5544
Epoch   0 Batch  370/538 - Train Accuracy: 0.6172, Validation Accuracy: 0.6499, Loss: 0.5856
Epoch   0 Batch  371/538 - Train Accuracy: 0.6302, Validation Accuracy

Epoch   0 Batch  451/538 - Train Accuracy: 0.6498, Validation Accuracy: 0.6705, Loss: 0.4971
Epoch   0 Batch  452/538 - Train Accuracy: 0.6648, Validation Accuracy: 0.6678, Loss: 0.4815
Epoch   0 Batch  453/538 - Train Accuracy: 0.6434, Validation Accuracy: 0.6609, Loss: 0.5083
Epoch   0 Batch  454/538 - Train Accuracy: 0.6687, Validation Accuracy: 0.6657, Loss: 0.4743
Epoch   0 Batch  455/538 - Train Accuracy: 0.6992, Validation Accuracy: 0.6626, Loss: 0.4500
Epoch   0 Batch  456/538 - Train Accuracy: 0.7183, Validation Accuracy: 0.6671, Loss: 0.4398
Epoch   0 Batch  457/538 - Train Accuracy: 0.6557, Validation Accuracy: 0.6577, Loss: 0.5038
Epoch   0 Batch  458/538 - Train Accuracy: 0.6695, Validation Accuracy: 0.6584, Loss: 0.4681
Epoch   0 Batch  459/538 - Train Accuracy: 0.6717, Validation Accuracy: 0.6674, Loss: 0.4847
Epoch   0 Batch  460/538 - Train Accuracy: 0.6551, Validation Accuracy: 0.6580, Loss: 0.4751
Epoch   0 Batch  461/538 - Train Accuracy: 0.6375, Validation Accuracy

Epoch   1 Batch    5/538 - Train Accuracy: 0.6687, Validation Accuracy: 0.6799, Loss: 0.4271
Epoch   1 Batch    6/538 - Train Accuracy: 0.7178, Validation Accuracy: 0.6543, Loss: 0.4053
Epoch   1 Batch    7/538 - Train Accuracy: 0.7004, Validation Accuracy: 0.6896, Loss: 0.4088
Epoch   1 Batch    8/538 - Train Accuracy: 0.7004, Validation Accuracy: 0.6799, Loss: 0.4227
Epoch   1 Batch    9/538 - Train Accuracy: 0.6971, Validation Accuracy: 0.6747, Loss: 0.4175
Epoch   1 Batch   10/538 - Train Accuracy: 0.6643, Validation Accuracy: 0.6710, Loss: 0.4466
Epoch   1 Batch   11/538 - Train Accuracy: 0.6891, Validation Accuracy: 0.6861, Loss: 0.4252
Epoch   1 Batch   12/538 - Train Accuracy: 0.6633, Validation Accuracy: 0.6774, Loss: 0.4148
Epoch   1 Batch   13/538 - Train Accuracy: 0.7054, Validation Accuracy: 0.6768, Loss: 0.3959
Epoch   1 Batch   14/538 - Train Accuracy: 0.6953, Validation Accuracy: 0.6925, Loss: 0.4069
Epoch   1 Batch   15/538 - Train Accuracy: 0.7108, Validation Accuracy

Epoch   1 Batch   95/538 - Train Accuracy: 0.7843, Validation Accuracy: 0.7670, Loss: 0.2912
Epoch   1 Batch   96/538 - Train Accuracy: 0.8039, Validation Accuracy: 0.7786, Loss: 0.2860
Epoch   1 Batch   97/538 - Train Accuracy: 0.7506, Validation Accuracy: 0.7619, Loss: 0.3023
Epoch   1 Batch   98/538 - Train Accuracy: 0.7954, Validation Accuracy: 0.7692, Loss: 0.2990
Epoch   1 Batch   99/538 - Train Accuracy: 0.7553, Validation Accuracy: 0.7797, Loss: 0.2986
Epoch   1 Batch  100/538 - Train Accuracy: 0.7730, Validation Accuracy: 0.7598, Loss: 0.2939
Epoch   1 Batch  101/538 - Train Accuracy: 0.7252, Validation Accuracy: 0.7713, Loss: 0.3124
Epoch   1 Batch  102/538 - Train Accuracy: 0.7824, Validation Accuracy: 0.7809, Loss: 0.3033
Epoch   1 Batch  103/538 - Train Accuracy: 0.7852, Validation Accuracy: 0.7756, Loss: 0.3020
Epoch   1 Batch  104/538 - Train Accuracy: 0.7571, Validation Accuracy: 0.7630, Loss: 0.2909
Epoch   1 Batch  105/538 - Train Accuracy: 0.7673, Validation Accuracy

Epoch   1 Batch  184/538 - Train Accuracy: 0.8110, Validation Accuracy: 0.8240, Loss: 0.2156
Epoch   1 Batch  185/538 - Train Accuracy: 0.8416, Validation Accuracy: 0.8152, Loss: 0.2153
Epoch   1 Batch  186/538 - Train Accuracy: 0.8583, Validation Accuracy: 0.8148, Loss: 0.2229
Epoch   1 Batch  187/538 - Train Accuracy: 0.8208, Validation Accuracy: 0.8152, Loss: 0.2090
Epoch   1 Batch  188/538 - Train Accuracy: 0.8256, Validation Accuracy: 0.8185, Loss: 0.2146
Epoch   1 Batch  189/538 - Train Accuracy: 0.8168, Validation Accuracy: 0.8159, Loss: 0.2275
Epoch   1 Batch  190/538 - Train Accuracy: 0.8002, Validation Accuracy: 0.8153, Loss: 0.2496
Epoch   1 Batch  191/538 - Train Accuracy: 0.8265, Validation Accuracy: 0.8237, Loss: 0.2063
Epoch   1 Batch  192/538 - Train Accuracy: 0.8348, Validation Accuracy: 0.8253, Loss: 0.2091
Epoch   1 Batch  193/538 - Train Accuracy: 0.8259, Validation Accuracy: 0.8317, Loss: 0.2072
Epoch   1 Batch  194/538 - Train Accuracy: 0.8068, Validation Accuracy

Epoch   1 Batch  273/538 - Train Accuracy: 0.8793, Validation Accuracy: 0.8468, Loss: 0.1567
Epoch   1 Batch  274/538 - Train Accuracy: 0.8268, Validation Accuracy: 0.8512, Loss: 0.1693
Epoch   1 Batch  275/538 - Train Accuracy: 0.8832, Validation Accuracy: 0.8608, Loss: 0.1632
Epoch   1 Batch  276/538 - Train Accuracy: 0.8467, Validation Accuracy: 0.8519, Loss: 0.1616
Epoch   1 Batch  277/538 - Train Accuracy: 0.8748, Validation Accuracy: 0.8475, Loss: 0.1435
Epoch   1 Batch  278/538 - Train Accuracy: 0.8752, Validation Accuracy: 0.8485, Loss: 0.1490
Epoch   1 Batch  279/538 - Train Accuracy: 0.8568, Validation Accuracy: 0.8610, Loss: 0.1521
Epoch   1 Batch  280/538 - Train Accuracy: 0.8631, Validation Accuracy: 0.8331, Loss: 0.1410
Epoch   1 Batch  281/538 - Train Accuracy: 0.8695, Validation Accuracy: 0.8420, Loss: 0.1686
Epoch   1 Batch  282/538 - Train Accuracy: 0.8690, Validation Accuracy: 0.8411, Loss: 0.1619
Epoch   1 Batch  283/538 - Train Accuracy: 0.8580, Validation Accuracy

Epoch   1 Batch  362/538 - Train Accuracy: 0.8953, Validation Accuracy: 0.8702, Loss: 0.1058
Epoch   1 Batch  363/538 - Train Accuracy: 0.8782, Validation Accuracy: 0.8691, Loss: 0.1156
Epoch   1 Batch  364/538 - Train Accuracy: 0.9033, Validation Accuracy: 0.8787, Loss: 0.1365
Epoch   1 Batch  365/538 - Train Accuracy: 0.8731, Validation Accuracy: 0.8816, Loss: 0.1208
Epoch   1 Batch  366/538 - Train Accuracy: 0.9158, Validation Accuracy: 0.8842, Loss: 0.1140
Epoch   1 Batch  367/538 - Train Accuracy: 0.9029, Validation Accuracy: 0.8798, Loss: 0.1025
Epoch   1 Batch  368/538 - Train Accuracy: 0.8962, Validation Accuracy: 0.8910, Loss: 0.1105
Epoch   1 Batch  369/538 - Train Accuracy: 0.8953, Validation Accuracy: 0.8805, Loss: 0.1103
Epoch   1 Batch  370/538 - Train Accuracy: 0.8811, Validation Accuracy: 0.8704, Loss: 0.1145
Epoch   1 Batch  371/538 - Train Accuracy: 0.9081, Validation Accuracy: 0.8622, Loss: 0.1096
Epoch   1 Batch  372/538 - Train Accuracy: 0.9039, Validation Accuracy

Epoch   1 Batch  452/538 - Train Accuracy: 0.9021, Validation Accuracy: 0.8901, Loss: 0.0878
Epoch   1 Batch  453/538 - Train Accuracy: 0.9180, Validation Accuracy: 0.8885, Loss: 0.1060
Epoch   1 Batch  454/538 - Train Accuracy: 0.8992, Validation Accuracy: 0.8876, Loss: 0.0964
Epoch   1 Batch  455/538 - Train Accuracy: 0.9013, Validation Accuracy: 0.8773, Loss: 0.0913
Epoch   1 Batch  456/538 - Train Accuracy: 0.9268, Validation Accuracy: 0.8736, Loss: 0.1112
Epoch   1 Batch  457/538 - Train Accuracy: 0.8887, Validation Accuracy: 0.8745, Loss: 0.0994
Epoch   1 Batch  458/538 - Train Accuracy: 0.9033, Validation Accuracy: 0.8762, Loss: 0.0822
Epoch   1 Batch  459/538 - Train Accuracy: 0.9057, Validation Accuracy: 0.8716, Loss: 0.0924
Epoch   1 Batch  460/538 - Train Accuracy: 0.8821, Validation Accuracy: 0.8849, Loss: 0.1020
Epoch   1 Batch  461/538 - Train Accuracy: 0.9116, Validation Accuracy: 0.8910, Loss: 0.0954
Epoch   1 Batch  462/538 - Train Accuracy: 0.8968, Validation Accuracy

Epoch   2 Batch    5/538 - Train Accuracy: 0.9087, Validation Accuracy: 0.9018, Loss: 0.0842
Epoch   2 Batch    6/538 - Train Accuracy: 0.9241, Validation Accuracy: 0.9023, Loss: 0.0714
Epoch   2 Batch    7/538 - Train Accuracy: 0.9301, Validation Accuracy: 0.9157, Loss: 0.0822
Epoch   2 Batch    8/538 - Train Accuracy: 0.9064, Validation Accuracy: 0.9171, Loss: 0.0781
Epoch   2 Batch    9/538 - Train Accuracy: 0.9088, Validation Accuracy: 0.9025, Loss: 0.0852
Epoch   2 Batch   10/538 - Train Accuracy: 0.9121, Validation Accuracy: 0.8936, Loss: 0.0865
Epoch   2 Batch   11/538 - Train Accuracy: 0.9131, Validation Accuracy: 0.8935, Loss: 0.0918
Epoch   2 Batch   12/538 - Train Accuracy: 0.9182, Validation Accuracy: 0.9068, Loss: 0.0989
Epoch   2 Batch   13/538 - Train Accuracy: 0.9243, Validation Accuracy: 0.9087, Loss: 0.0743
Epoch   2 Batch   14/538 - Train Accuracy: 0.9320, Validation Accuracy: 0.9023, Loss: 0.0794
Epoch   2 Batch   15/538 - Train Accuracy: 0.9172, Validation Accuracy

Epoch   2 Batch   94/538 - Train Accuracy: 0.9045, Validation Accuracy: 0.8828, Loss: 0.0636
Epoch   2 Batch   95/538 - Train Accuracy: 0.9153, Validation Accuracy: 0.9125, Loss: 0.0707
Epoch   2 Batch   96/538 - Train Accuracy: 0.9394, Validation Accuracy: 0.9103, Loss: 0.0582
Epoch   2 Batch   97/538 - Train Accuracy: 0.9207, Validation Accuracy: 0.9057, Loss: 0.0722
Epoch   2 Batch   98/538 - Train Accuracy: 0.9165, Validation Accuracy: 0.8880, Loss: 0.0741
Epoch   2 Batch   99/538 - Train Accuracy: 0.9113, Validation Accuracy: 0.9091, Loss: 0.0725
Epoch   2 Batch  100/538 - Train Accuracy: 0.9221, Validation Accuracy: 0.9119, Loss: 0.0599
Epoch   2 Batch  101/538 - Train Accuracy: 0.8947, Validation Accuracy: 0.9096, Loss: 0.0846
Epoch   2 Batch  102/538 - Train Accuracy: 0.9150, Validation Accuracy: 0.9031, Loss: 0.0787
Epoch   2 Batch  103/538 - Train Accuracy: 0.9297, Validation Accuracy: 0.8961, Loss: 0.0725
Epoch   2 Batch  104/538 - Train Accuracy: 0.9133, Validation Accuracy

Epoch   2 Batch  184/538 - Train Accuracy: 0.9308, Validation Accuracy: 0.9171, Loss: 0.0633
Epoch   2 Batch  185/538 - Train Accuracy: 0.9453, Validation Accuracy: 0.9247, Loss: 0.0596
Epoch   2 Batch  186/538 - Train Accuracy: 0.9144, Validation Accuracy: 0.9144, Loss: 0.0634
Epoch   2 Batch  187/538 - Train Accuracy: 0.9208, Validation Accuracy: 0.9091, Loss: 0.0668
Epoch   2 Batch  188/538 - Train Accuracy: 0.9189, Validation Accuracy: 0.9078, Loss: 0.0600
Epoch   2 Batch  189/538 - Train Accuracy: 0.9299, Validation Accuracy: 0.9080, Loss: 0.0685
Epoch   2 Batch  190/538 - Train Accuracy: 0.9003, Validation Accuracy: 0.9272, Loss: 0.0915
Epoch   2 Batch  191/538 - Train Accuracy: 0.9301, Validation Accuracy: 0.9158, Loss: 0.0630
Epoch   2 Batch  192/538 - Train Accuracy: 0.9208, Validation Accuracy: 0.9178, Loss: 0.0629
Epoch   2 Batch  193/538 - Train Accuracy: 0.9245, Validation Accuracy: 0.8972, Loss: 0.0618
Epoch   2 Batch  194/538 - Train Accuracy: 0.8945, Validation Accuracy

Epoch   2 Batch  274/538 - Train Accuracy: 0.8975, Validation Accuracy: 0.9123, Loss: 0.0627
Epoch   2 Batch  275/538 - Train Accuracy: 0.9199, Validation Accuracy: 0.8995, Loss: 0.0653
Epoch   2 Batch  276/538 - Train Accuracy: 0.9158, Validation Accuracy: 0.9043, Loss: 0.0636
Epoch   2 Batch  277/538 - Train Accuracy: 0.9357, Validation Accuracy: 0.9183, Loss: 0.0516
Epoch   2 Batch  278/538 - Train Accuracy: 0.9086, Validation Accuracy: 0.9245, Loss: 0.0559
Epoch   2 Batch  279/538 - Train Accuracy: 0.9258, Validation Accuracy: 0.9254, Loss: 0.0592
Epoch   2 Batch  280/538 - Train Accuracy: 0.9340, Validation Accuracy: 0.9237, Loss: 0.0513
Epoch   2 Batch  281/538 - Train Accuracy: 0.9295, Validation Accuracy: 0.9293, Loss: 0.0575
Epoch   2 Batch  282/538 - Train Accuracy: 0.9198, Validation Accuracy: 0.9142, Loss: 0.0672
Epoch   2 Batch  283/538 - Train Accuracy: 0.9418, Validation Accuracy: 0.9208, Loss: 0.0588
Epoch   2 Batch  284/538 - Train Accuracy: 0.9383, Validation Accuracy

Epoch   2 Batch  364/538 - Train Accuracy: 0.9287, Validation Accuracy: 0.9185, Loss: 0.0785
Epoch   2 Batch  365/538 - Train Accuracy: 0.9126, Validation Accuracy: 0.9077, Loss: 0.0702
Epoch   2 Batch  366/538 - Train Accuracy: 0.9406, Validation Accuracy: 0.9077, Loss: 0.0613
Epoch   2 Batch  367/538 - Train Accuracy: 0.9281, Validation Accuracy: 0.9034, Loss: 0.0561
Epoch   2 Batch  368/538 - Train Accuracy: 0.9209, Validation Accuracy: 0.9119, Loss: 0.0553
Epoch   2 Batch  369/538 - Train Accuracy: 0.9334, Validation Accuracy: 0.9306, Loss: 0.0557
Epoch   2 Batch  370/538 - Train Accuracy: 0.9437, Validation Accuracy: 0.9361, Loss: 0.0598
Epoch   2 Batch  371/538 - Train Accuracy: 0.9481, Validation Accuracy: 0.9276, Loss: 0.0541
Epoch   2 Batch  372/538 - Train Accuracy: 0.9488, Validation Accuracy: 0.9247, Loss: 0.0532
Epoch   2 Batch  373/538 - Train Accuracy: 0.9236, Validation Accuracy: 0.9251, Loss: 0.0501
Epoch   2 Batch  374/538 - Train Accuracy: 0.9242, Validation Accuracy

Epoch   2 Batch  454/538 - Train Accuracy: 0.9351, Validation Accuracy: 0.9450, Loss: 0.0505
Epoch   2 Batch  455/538 - Train Accuracy: 0.9455, Validation Accuracy: 0.9517, Loss: 0.0488
Epoch   2 Batch  456/538 - Train Accuracy: 0.9447, Validation Accuracy: 0.9558, Loss: 0.0627
Epoch   2 Batch  457/538 - Train Accuracy: 0.9381, Validation Accuracy: 0.9522, Loss: 0.0527
Epoch   2 Batch  458/538 - Train Accuracy: 0.9611, Validation Accuracy: 0.9444, Loss: 0.0421
Epoch   2 Batch  459/538 - Train Accuracy: 0.9455, Validation Accuracy: 0.9510, Loss: 0.0448
Epoch   2 Batch  460/538 - Train Accuracy: 0.9263, Validation Accuracy: 0.9473, Loss: 0.0503
Epoch   2 Batch  461/538 - Train Accuracy: 0.9480, Validation Accuracy: 0.9432, Loss: 0.0569
Epoch   2 Batch  462/538 - Train Accuracy: 0.9501, Validation Accuracy: 0.9485, Loss: 0.0451
Epoch   2 Batch  463/538 - Train Accuracy: 0.9295, Validation Accuracy: 0.9451, Loss: 0.0484
Epoch   2 Batch  464/538 - Train Accuracy: 0.9303, Validation Accuracy

Epoch   3 Batch    7/538 - Train Accuracy: 0.9434, Validation Accuracy: 0.9240, Loss: 0.0490
Epoch   3 Batch    8/538 - Train Accuracy: 0.9428, Validation Accuracy: 0.9254, Loss: 0.0495
Epoch   3 Batch    9/538 - Train Accuracy: 0.9342, Validation Accuracy: 0.9215, Loss: 0.0442
Epoch   3 Batch   10/538 - Train Accuracy: 0.9412, Validation Accuracy: 0.9224, Loss: 0.0465
Epoch   3 Batch   11/538 - Train Accuracy: 0.9463, Validation Accuracy: 0.9261, Loss: 0.0435
Epoch   3 Batch   12/538 - Train Accuracy: 0.9477, Validation Accuracy: 0.9256, Loss: 0.0449
Epoch   3 Batch   13/538 - Train Accuracy: 0.9555, Validation Accuracy: 0.9386, Loss: 0.0418
Epoch   3 Batch   14/538 - Train Accuracy: 0.9502, Validation Accuracy: 0.9366, Loss: 0.0413
Epoch   3 Batch   15/538 - Train Accuracy: 0.9488, Validation Accuracy: 0.9341, Loss: 0.0421
Epoch   3 Batch   16/538 - Train Accuracy: 0.9388, Validation Accuracy: 0.9336, Loss: 0.0433
Epoch   3 Batch   17/538 - Train Accuracy: 0.9584, Validation Accuracy

Epoch   3 Batch   97/538 - Train Accuracy: 0.9561, Validation Accuracy: 0.9384, Loss: 0.0342
Epoch   3 Batch   98/538 - Train Accuracy: 0.9364, Validation Accuracy: 0.9425, Loss: 0.0438
Epoch   3 Batch   99/538 - Train Accuracy: 0.9582, Validation Accuracy: 0.9384, Loss: 0.0345
Epoch   3 Batch  100/538 - Train Accuracy: 0.9563, Validation Accuracy: 0.9391, Loss: 0.0363
Epoch   3 Batch  101/538 - Train Accuracy: 0.9273, Validation Accuracy: 0.9316, Loss: 0.0498
Epoch   3 Batch  102/538 - Train Accuracy: 0.9432, Validation Accuracy: 0.9492, Loss: 0.0413
Epoch   3 Batch  103/538 - Train Accuracy: 0.9464, Validation Accuracy: 0.9419, Loss: 0.0391
Epoch   3 Batch  104/538 - Train Accuracy: 0.9505, Validation Accuracy: 0.9395, Loss: 0.0361
Epoch   3 Batch  105/538 - Train Accuracy: 0.9481, Validation Accuracy: 0.9334, Loss: 0.0377
Epoch   3 Batch  106/538 - Train Accuracy: 0.9264, Validation Accuracy: 0.9267, Loss: 0.0379
Epoch   3 Batch  107/538 - Train Accuracy: 0.9340, Validation Accuracy

Epoch   3 Batch  186/538 - Train Accuracy: 0.9570, Validation Accuracy: 0.9293, Loss: 0.0399
Epoch   3 Batch  187/538 - Train Accuracy: 0.9583, Validation Accuracy: 0.9244, Loss: 0.0394
Epoch   3 Batch  188/538 - Train Accuracy: 0.9457, Validation Accuracy: 0.9261, Loss: 0.0355
Epoch   3 Batch  189/538 - Train Accuracy: 0.9520, Validation Accuracy: 0.9260, Loss: 0.0418
Epoch   3 Batch  190/538 - Train Accuracy: 0.9347, Validation Accuracy: 0.9322, Loss: 0.0552
Epoch   3 Batch  191/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9498, Loss: 0.0373
Epoch   3 Batch  192/538 - Train Accuracy: 0.9416, Validation Accuracy: 0.9439, Loss: 0.0384
Epoch   3 Batch  193/538 - Train Accuracy: 0.9533, Validation Accuracy: 0.9357, Loss: 0.0390
Epoch   3 Batch  194/538 - Train Accuracy: 0.9289, Validation Accuracy: 0.9235, Loss: 0.0473
Epoch   3 Batch  195/538 - Train Accuracy: 0.9528, Validation Accuracy: 0.9251, Loss: 0.0419
Epoch   3 Batch  196/538 - Train Accuracy: 0.9395, Validation Accuracy

Epoch   3 Batch  275/538 - Train Accuracy: 0.9566, Validation Accuracy: 0.9434, Loss: 0.0404
Epoch   3 Batch  276/538 - Train Accuracy: 0.9389, Validation Accuracy: 0.9490, Loss: 0.0407
Epoch   3 Batch  277/538 - Train Accuracy: 0.9473, Validation Accuracy: 0.9492, Loss: 0.0298
Epoch   3 Batch  278/538 - Train Accuracy: 0.9453, Validation Accuracy: 0.9540, Loss: 0.0354
Epoch   3 Batch  279/538 - Train Accuracy: 0.9473, Validation Accuracy: 0.9531, Loss: 0.0370
Epoch   3 Batch  280/538 - Train Accuracy: 0.9555, Validation Accuracy: 0.9581, Loss: 0.0325
Epoch   3 Batch  281/538 - Train Accuracy: 0.9400, Validation Accuracy: 0.9613, Loss: 0.0408
Epoch   3 Batch  282/538 - Train Accuracy: 0.9427, Validation Accuracy: 0.9590, Loss: 0.0395
Epoch   3 Batch  283/538 - Train Accuracy: 0.9611, Validation Accuracy: 0.9581, Loss: 0.0376
Epoch   3 Batch  284/538 - Train Accuracy: 0.9416, Validation Accuracy: 0.9585, Loss: 0.0409
Epoch   3 Batch  285/538 - Train Accuracy: 0.9544, Validation Accuracy

Epoch   3 Batch  365/538 - Train Accuracy: 0.9524, Validation Accuracy: 0.9631, Loss: 0.0360
Epoch   3 Batch  366/538 - Train Accuracy: 0.9600, Validation Accuracy: 0.9680, Loss: 0.0370
Epoch   3 Batch  367/538 - Train Accuracy: 0.9641, Validation Accuracy: 0.9556, Loss: 0.0309
Epoch   3 Batch  368/538 - Train Accuracy: 0.9494, Validation Accuracy: 0.9460, Loss: 0.0311
Epoch   3 Batch  369/538 - Train Accuracy: 0.9535, Validation Accuracy: 0.9501, Loss: 0.0319
Epoch   3 Batch  370/538 - Train Accuracy: 0.9590, Validation Accuracy: 0.9506, Loss: 0.0290
Epoch   3 Batch  371/538 - Train Accuracy: 0.9619, Validation Accuracy: 0.9480, Loss: 0.0364
Epoch   3 Batch  372/538 - Train Accuracy: 0.9646, Validation Accuracy: 0.9423, Loss: 0.0324
Epoch   3 Batch  373/538 - Train Accuracy: 0.9520, Validation Accuracy: 0.9487, Loss: 0.0296
Epoch   3 Batch  374/538 - Train Accuracy: 0.9523, Validation Accuracy: 0.9478, Loss: 0.0357
Epoch   3 Batch  375/538 - Train Accuracy: 0.9475, Validation Accuracy

Epoch   3 Batch  454/538 - Train Accuracy: 0.9520, Validation Accuracy: 0.9537, Loss: 0.0368
Epoch   3 Batch  455/538 - Train Accuracy: 0.9647, Validation Accuracy: 0.9585, Loss: 0.0289
Epoch   3 Batch  456/538 - Train Accuracy: 0.9606, Validation Accuracy: 0.9638, Loss: 0.0463
Epoch   3 Batch  457/538 - Train Accuracy: 0.9574, Validation Accuracy: 0.9652, Loss: 0.0333
Epoch   3 Batch  458/538 - Train Accuracy: 0.9680, Validation Accuracy: 0.9645, Loss: 0.0281
Epoch   3 Batch  459/538 - Train Accuracy: 0.9699, Validation Accuracy: 0.9613, Loss: 0.0284
Epoch   3 Batch  460/538 - Train Accuracy: 0.9453, Validation Accuracy: 0.9613, Loss: 0.0340
Epoch   3 Batch  461/538 - Train Accuracy: 0.9599, Validation Accuracy: 0.9590, Loss: 0.0330
Epoch   3 Batch  462/538 - Train Accuracy: 0.9600, Validation Accuracy: 0.9592, Loss: 0.0331
Epoch   3 Batch  463/538 - Train Accuracy: 0.9301, Validation Accuracy: 0.9622, Loss: 0.0362
Epoch   3 Batch  464/538 - Train Accuracy: 0.9568, Validation Accuracy

Epoch   4 Batch    8/538 - Train Accuracy: 0.9652, Validation Accuracy: 0.9638, Loss: 0.0325
Epoch   4 Batch    9/538 - Train Accuracy: 0.9598, Validation Accuracy: 0.9569, Loss: 0.0306
Epoch   4 Batch   10/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9551, Loss: 0.0333
Epoch   4 Batch   11/538 - Train Accuracy: 0.9613, Validation Accuracy: 0.9506, Loss: 0.0305
Epoch   4 Batch   12/538 - Train Accuracy: 0.9555, Validation Accuracy: 0.9537, Loss: 0.0324
Epoch   4 Batch   13/538 - Train Accuracy: 0.9712, Validation Accuracy: 0.9515, Loss: 0.0297
Epoch   4 Batch   14/538 - Train Accuracy: 0.9504, Validation Accuracy: 0.9522, Loss: 0.0295
Epoch   4 Batch   15/538 - Train Accuracy: 0.9604, Validation Accuracy: 0.9586, Loss: 0.0283
Epoch   4 Batch   16/538 - Train Accuracy: 0.9494, Validation Accuracy: 0.9529, Loss: 0.0371
Epoch   4 Batch   17/538 - Train Accuracy: 0.9746, Validation Accuracy: 0.9540, Loss: 0.0368
Epoch   4 Batch   18/538 - Train Accuracy: 0.9701, Validation Accuracy

Epoch   4 Batch   97/538 - Train Accuracy: 0.9646, Validation Accuracy: 0.9576, Loss: 0.0264
Epoch   4 Batch   98/538 - Train Accuracy: 0.9667, Validation Accuracy: 0.9600, Loss: 0.0300
Epoch   4 Batch   99/538 - Train Accuracy: 0.9674, Validation Accuracy: 0.9551, Loss: 0.0253
Epoch   4 Batch  100/538 - Train Accuracy: 0.9709, Validation Accuracy: 0.9538, Loss: 0.0253
Epoch   4 Batch  101/538 - Train Accuracy: 0.9512, Validation Accuracy: 0.9524, Loss: 0.0373
Epoch   4 Batch  102/538 - Train Accuracy: 0.9609, Validation Accuracy: 0.9498, Loss: 0.0326
Epoch   4 Batch  103/538 - Train Accuracy: 0.9585, Validation Accuracy: 0.9498, Loss: 0.0308
Epoch   4 Batch  104/538 - Train Accuracy: 0.9747, Validation Accuracy: 0.9540, Loss: 0.0284
Epoch   4 Batch  105/538 - Train Accuracy: 0.9617, Validation Accuracy: 0.9595, Loss: 0.0236
Epoch   4 Batch  106/538 - Train Accuracy: 0.9426, Validation Accuracy: 0.9565, Loss: 0.0261
Epoch   4 Batch  107/538 - Train Accuracy: 0.9578, Validation Accuracy

Epoch   4 Batch  187/538 - Train Accuracy: 0.9716, Validation Accuracy: 0.9583, Loss: 0.0257
Epoch   4 Batch  188/538 - Train Accuracy: 0.9584, Validation Accuracy: 0.9577, Loss: 0.0261
Epoch   4 Batch  189/538 - Train Accuracy: 0.9744, Validation Accuracy: 0.9515, Loss: 0.0284
Epoch   4 Batch  190/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9547, Loss: 0.0389
Epoch   4 Batch  191/538 - Train Accuracy: 0.9706, Validation Accuracy: 0.9592, Loss: 0.0283
Epoch   4 Batch  192/538 - Train Accuracy: 0.9624, Validation Accuracy: 0.9597, Loss: 0.0285
Epoch   4 Batch  193/538 - Train Accuracy: 0.9624, Validation Accuracy: 0.9572, Loss: 0.0320
Epoch   4 Batch  194/538 - Train Accuracy: 0.9514, Validation Accuracy: 0.9616, Loss: 0.0341
Epoch   4 Batch  195/538 - Train Accuracy: 0.9689, Validation Accuracy: 0.9554, Loss: 0.0330
Epoch   4 Batch  196/538 - Train Accuracy: 0.9516, Validation Accuracy: 0.9544, Loss: 0.0277
Epoch   4 Batch  197/538 - Train Accuracy: 0.9611, Validation Accuracy

Epoch   4 Batch  277/538 - Train Accuracy: 0.9705, Validation Accuracy: 0.9569, Loss: 0.0222
Epoch   4 Batch  278/538 - Train Accuracy: 0.9604, Validation Accuracy: 0.9675, Loss: 0.0233
Epoch   4 Batch  279/538 - Train Accuracy: 0.9563, Validation Accuracy: 0.9650, Loss: 0.0316
Epoch   4 Batch  280/538 - Train Accuracy: 0.9674, Validation Accuracy: 0.9703, Loss: 0.0227
Epoch   4 Batch  281/538 - Train Accuracy: 0.9482, Validation Accuracy: 0.9664, Loss: 0.0290
Epoch   4 Batch  282/538 - Train Accuracy: 0.9652, Validation Accuracy: 0.9643, Loss: 0.0280
Epoch   4 Batch  283/538 - Train Accuracy: 0.9734, Validation Accuracy: 0.9656, Loss: 0.0243
Epoch   4 Batch  284/538 - Train Accuracy: 0.9584, Validation Accuracy: 0.9600, Loss: 0.0307
Epoch   4 Batch  285/538 - Train Accuracy: 0.9628, Validation Accuracy: 0.9544, Loss: 0.0307
Epoch   4 Batch  286/538 - Train Accuracy: 0.9543, Validation Accuracy: 0.9513, Loss: 0.0370
Epoch   4 Batch  287/538 - Train Accuracy: 0.9621, Validation Accuracy

Epoch   4 Batch  367/538 - Train Accuracy: 0.9689, Validation Accuracy: 0.9711, Loss: 0.0205
Epoch   4 Batch  368/538 - Train Accuracy: 0.9704, Validation Accuracy: 0.9723, Loss: 0.0243
Epoch   4 Batch  369/538 - Train Accuracy: 0.9602, Validation Accuracy: 0.9735, Loss: 0.0258
Epoch   4 Batch  370/538 - Train Accuracy: 0.9715, Validation Accuracy: 0.9675, Loss: 0.0232
Epoch   4 Batch  371/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9604, Loss: 0.0271
Epoch   4 Batch  372/538 - Train Accuracy: 0.9672, Validation Accuracy: 0.9551, Loss: 0.0261
Epoch   4 Batch  373/538 - Train Accuracy: 0.9570, Validation Accuracy: 0.9501, Loss: 0.0262
Epoch   4 Batch  374/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9545, Loss: 0.0228
Epoch   4 Batch  375/538 - Train Accuracy: 0.9585, Validation Accuracy: 0.9629, Loss: 0.0271
Epoch   4 Batch  376/538 - Train Accuracy: 0.9529, Validation Accuracy: 0.9679, Loss: 0.0294
Epoch   4 Batch  377/538 - Train Accuracy: 0.9684, Validation Accuracy

Epoch   4 Batch  456/538 - Train Accuracy: 0.9601, Validation Accuracy: 0.9689, Loss: 0.0423
Epoch   4 Batch  457/538 - Train Accuracy: 0.9682, Validation Accuracy: 0.9712, Loss: 0.0294
Epoch   4 Batch  458/538 - Train Accuracy: 0.9684, Validation Accuracy: 0.9691, Loss: 0.0255
Epoch   4 Batch  459/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9698, Loss: 0.0241
Epoch   4 Batch  460/538 - Train Accuracy: 0.9581, Validation Accuracy: 0.9670, Loss: 0.0297
Epoch   4 Batch  461/538 - Train Accuracy: 0.9698, Validation Accuracy: 0.9645, Loss: 0.0324
Epoch   4 Batch  462/538 - Train Accuracy: 0.9635, Validation Accuracy: 0.9604, Loss: 0.0257
Epoch   4 Batch  463/538 - Train Accuracy: 0.9527, Validation Accuracy: 0.9570, Loss: 0.0356
Epoch   4 Batch  464/538 - Train Accuracy: 0.9680, Validation Accuracy: 0.9551, Loss: 0.0254
Epoch   4 Batch  465/538 - Train Accuracy: 0.9678, Validation Accuracy: 0.9640, Loss: 0.0296
Epoch   4 Batch  466/538 - Train Accuracy: 0.9498, Validation Accuracy

Epoch   5 Batch   10/538 - Train Accuracy: 0.9637, Validation Accuracy: 0.9718, Loss: 0.0220
Epoch   5 Batch   11/538 - Train Accuracy: 0.9758, Validation Accuracy: 0.9672, Loss: 0.0285
Epoch   5 Batch   12/538 - Train Accuracy: 0.9674, Validation Accuracy: 0.9696, Loss: 0.0235
Epoch   5 Batch   13/538 - Train Accuracy: 0.9667, Validation Accuracy: 0.9602, Loss: 0.0235
Epoch   5 Batch   14/538 - Train Accuracy: 0.9719, Validation Accuracy: 0.9680, Loss: 0.0193
Epoch   5 Batch   15/538 - Train Accuracy: 0.9619, Validation Accuracy: 0.9650, Loss: 0.0256
Epoch   5 Batch   16/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9622, Loss: 0.0247
Epoch   5 Batch   17/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9707, Loss: 0.0264
Epoch   5 Batch   18/538 - Train Accuracy: 0.9775, Validation Accuracy: 0.9677, Loss: 0.0292
Epoch   5 Batch   19/538 - Train Accuracy: 0.9693, Validation Accuracy: 0.9668, Loss: 0.0260
Epoch   5 Batch   20/538 - Train Accuracy: 0.9667, Validation Accuracy

Epoch   5 Batch  100/538 - Train Accuracy: 0.9773, Validation Accuracy: 0.9719, Loss: 0.0217
Epoch   5 Batch  101/538 - Train Accuracy: 0.9752, Validation Accuracy: 0.9641, Loss: 0.0297
Epoch   5 Batch  102/538 - Train Accuracy: 0.9656, Validation Accuracy: 0.9641, Loss: 0.0279
Epoch   5 Batch  103/538 - Train Accuracy: 0.9678, Validation Accuracy: 0.9641, Loss: 0.0272
Epoch   5 Batch  104/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9654, Loss: 0.0203
Epoch   5 Batch  105/538 - Train Accuracy: 0.9626, Validation Accuracy: 0.9672, Loss: 0.0214
Epoch   5 Batch  106/538 - Train Accuracy: 0.9516, Validation Accuracy: 0.9572, Loss: 0.0236
Epoch   5 Batch  107/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9632, Loss: 0.0271
Epoch   5 Batch  108/538 - Train Accuracy: 0.9693, Validation Accuracy: 0.9627, Loss: 0.0240
Epoch   5 Batch  109/538 - Train Accuracy: 0.9713, Validation Accuracy: 0.9647, Loss: 0.0232
Epoch   5 Batch  110/538 - Train Accuracy: 0.9768, Validation Accuracy

Epoch   5 Batch  189/538 - Train Accuracy: 0.9699, Validation Accuracy: 0.9592, Loss: 0.0267
Epoch   5 Batch  190/538 - Train Accuracy: 0.9565, Validation Accuracy: 0.9593, Loss: 0.0299
Epoch   5 Batch  191/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9590, Loss: 0.0209
Epoch   5 Batch  192/538 - Train Accuracy: 0.9738, Validation Accuracy: 0.9590, Loss: 0.0184
Epoch   5 Batch  193/538 - Train Accuracy: 0.9745, Validation Accuracy: 0.9540, Loss: 0.0232
Epoch   5 Batch  194/538 - Train Accuracy: 0.9563, Validation Accuracy: 0.9529, Loss: 0.0242
Epoch   5 Batch  195/538 - Train Accuracy: 0.9740, Validation Accuracy: 0.9524, Loss: 0.0287
Epoch   5 Batch  196/538 - Train Accuracy: 0.9758, Validation Accuracy: 0.9508, Loss: 0.0220
Epoch   5 Batch  197/538 - Train Accuracy: 0.9619, Validation Accuracy: 0.9480, Loss: 0.0245
Epoch   5 Batch  198/538 - Train Accuracy: 0.9689, Validation Accuracy: 0.9521, Loss: 0.0246
Epoch   5 Batch  199/538 - Train Accuracy: 0.9535, Validation Accuracy

Epoch   5 Batch  278/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9624, Loss: 0.0227
Epoch   5 Batch  279/538 - Train Accuracy: 0.9674, Validation Accuracy: 0.9602, Loss: 0.0199
Epoch   5 Batch  280/538 - Train Accuracy: 0.9699, Validation Accuracy: 0.9654, Loss: 0.0190
Epoch   5 Batch  281/538 - Train Accuracy: 0.9662, Validation Accuracy: 0.9741, Loss: 0.0236
Epoch   5 Batch  282/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9716, Loss: 0.0232
Epoch   5 Batch  283/538 - Train Accuracy: 0.9752, Validation Accuracy: 0.9744, Loss: 0.0230
Epoch   5 Batch  284/538 - Train Accuracy: 0.9623, Validation Accuracy: 0.9744, Loss: 0.0253
Epoch   5 Batch  285/538 - Train Accuracy: 0.9836, Validation Accuracy: 0.9796, Loss: 0.0205
Epoch   5 Batch  286/538 - Train Accuracy: 0.9520, Validation Accuracy: 0.9727, Loss: 0.0263
Epoch   5 Batch  287/538 - Train Accuracy: 0.9760, Validation Accuracy: 0.9822, Loss: 0.0196
Epoch   5 Batch  288/538 - Train Accuracy: 0.9686, Validation Accuracy

Epoch   5 Batch  367/538 - Train Accuracy: 0.9723, Validation Accuracy: 0.9597, Loss: 0.0198
Epoch   5 Batch  368/538 - Train Accuracy: 0.9871, Validation Accuracy: 0.9631, Loss: 0.0173
Epoch   5 Batch  369/538 - Train Accuracy: 0.9625, Validation Accuracy: 0.9675, Loss: 0.0185
Epoch   5 Batch  370/538 - Train Accuracy: 0.9770, Validation Accuracy: 0.9643, Loss: 0.0193
Epoch   5 Batch  371/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9613, Loss: 0.0214
Epoch   5 Batch  372/538 - Train Accuracy: 0.9746, Validation Accuracy: 0.9609, Loss: 0.0210
Epoch   5 Batch  373/538 - Train Accuracy: 0.9682, Validation Accuracy: 0.9679, Loss: 0.0164
Epoch   5 Batch  374/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9618, Loss: 0.0191
Epoch   5 Batch  375/538 - Train Accuracy: 0.9708, Validation Accuracy: 0.9672, Loss: 0.0206
Epoch   5 Batch  376/538 - Train Accuracy: 0.9680, Validation Accuracy: 0.9668, Loss: 0.0205
Epoch   5 Batch  377/538 - Train Accuracy: 0.9666, Validation Accuracy

Epoch   5 Batch  456/538 - Train Accuracy: 0.9735, Validation Accuracy: 0.9586, Loss: 0.0312
Epoch   5 Batch  457/538 - Train Accuracy: 0.9668, Validation Accuracy: 0.9622, Loss: 0.0203
Epoch   5 Batch  458/538 - Train Accuracy: 0.9799, Validation Accuracy: 0.9702, Loss: 0.0189
Epoch   5 Batch  459/538 - Train Accuracy: 0.9783, Validation Accuracy: 0.9721, Loss: 0.0158
Epoch   5 Batch  460/538 - Train Accuracy: 0.9686, Validation Accuracy: 0.9695, Loss: 0.0184
Epoch   5 Batch  461/538 - Train Accuracy: 0.9786, Validation Accuracy: 0.9641, Loss: 0.0208
Epoch   5 Batch  462/538 - Train Accuracy: 0.9693, Validation Accuracy: 0.9663, Loss: 0.0164
Epoch   5 Batch  463/538 - Train Accuracy: 0.9551, Validation Accuracy: 0.9677, Loss: 0.0257
Epoch   5 Batch  464/538 - Train Accuracy: 0.9855, Validation Accuracy: 0.9686, Loss: 0.0180
Epoch   5 Batch  465/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9686, Loss: 0.0184
Epoch   5 Batch  466/538 - Train Accuracy: 0.9680, Validation Accuracy

Epoch   6 Batch    9/538 - Train Accuracy: 0.9625, Validation Accuracy: 0.9663, Loss: 0.0167
Epoch   6 Batch   10/538 - Train Accuracy: 0.9768, Validation Accuracy: 0.9664, Loss: 0.0175
Epoch   6 Batch   11/538 - Train Accuracy: 0.9848, Validation Accuracy: 0.9705, Loss: 0.0217
Epoch   6 Batch   12/538 - Train Accuracy: 0.9754, Validation Accuracy: 0.9755, Loss: 0.0213
Epoch   6 Batch   13/538 - Train Accuracy: 0.9779, Validation Accuracy: 0.9766, Loss: 0.0171
Epoch   6 Batch   14/538 - Train Accuracy: 0.9801, Validation Accuracy: 0.9721, Loss: 0.0198
Epoch   6 Batch   15/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9668, Loss: 0.0175
Epoch   6 Batch   16/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9638, Loss: 0.0203
Epoch   6 Batch   17/538 - Train Accuracy: 0.9854, Validation Accuracy: 0.9604, Loss: 0.0204
Epoch   6 Batch   18/538 - Train Accuracy: 0.9793, Validation Accuracy: 0.9553, Loss: 0.0268
Epoch   6 Batch   19/538 - Train Accuracy: 0.9742, Validation Accuracy

Epoch   6 Batch   99/538 - Train Accuracy: 0.9756, Validation Accuracy: 0.9705, Loss: 0.0174
Epoch   6 Batch  100/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9648, Loss: 0.0156
Epoch   6 Batch  101/538 - Train Accuracy: 0.9713, Validation Accuracy: 0.9680, Loss: 0.0280
Epoch   6 Batch  102/538 - Train Accuracy: 0.9623, Validation Accuracy: 0.9682, Loss: 0.0200
Epoch   6 Batch  103/538 - Train Accuracy: 0.9794, Validation Accuracy: 0.9657, Loss: 0.0176
Epoch   6 Batch  104/538 - Train Accuracy: 0.9779, Validation Accuracy: 0.9652, Loss: 0.0181
Epoch   6 Batch  105/538 - Train Accuracy: 0.9676, Validation Accuracy: 0.9638, Loss: 0.0148
Epoch   6 Batch  106/538 - Train Accuracy: 0.9570, Validation Accuracy: 0.9696, Loss: 0.0189
Epoch   6 Batch  107/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9631, Loss: 0.0257
Epoch   6 Batch  108/538 - Train Accuracy: 0.9814, Validation Accuracy: 0.9657, Loss: 0.0182
Epoch   6 Batch  109/538 - Train Accuracy: 0.9744, Validation Accuracy

Epoch   6 Batch  189/538 - Train Accuracy: 0.9793, Validation Accuracy: 0.9652, Loss: 0.0191
Epoch   6 Batch  190/538 - Train Accuracy: 0.9665, Validation Accuracy: 0.9632, Loss: 0.0225
Epoch   6 Batch  191/538 - Train Accuracy: 0.9860, Validation Accuracy: 0.9643, Loss: 0.0177
Epoch   6 Batch  192/538 - Train Accuracy: 0.9807, Validation Accuracy: 0.9636, Loss: 0.0156
Epoch   6 Batch  193/538 - Train Accuracy: 0.9836, Validation Accuracy: 0.9595, Loss: 0.0183
Epoch   6 Batch  194/538 - Train Accuracy: 0.9652, Validation Accuracy: 0.9588, Loss: 0.0184
Epoch   6 Batch  195/538 - Train Accuracy: 0.9769, Validation Accuracy: 0.9529, Loss: 0.0246
Epoch   6 Batch  196/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9522, Loss: 0.0177
Epoch   6 Batch  197/538 - Train Accuracy: 0.9795, Validation Accuracy: 0.9540, Loss: 0.0189
Epoch   6 Batch  198/538 - Train Accuracy: 0.9775, Validation Accuracy: 0.9513, Loss: 0.0201
Epoch   6 Batch  199/538 - Train Accuracy: 0.9664, Validation Accuracy

Epoch   6 Batch  279/538 - Train Accuracy: 0.9670, Validation Accuracy: 0.9680, Loss: 0.0171
Epoch   6 Batch  280/538 - Train Accuracy: 0.9786, Validation Accuracy: 0.9680, Loss: 0.0152
Epoch   6 Batch  281/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9686, Loss: 0.0195
Epoch   6 Batch  282/538 - Train Accuracy: 0.9810, Validation Accuracy: 0.9686, Loss: 0.0226
Epoch   6 Batch  283/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9702, Loss: 0.0202
Epoch   6 Batch  284/538 - Train Accuracy: 0.9760, Validation Accuracy: 0.9732, Loss: 0.0180
Epoch   6 Batch  285/538 - Train Accuracy: 0.9745, Validation Accuracy: 0.9776, Loss: 0.0178
Epoch   6 Batch  286/538 - Train Accuracy: 0.9719, Validation Accuracy: 0.9753, Loss: 0.0242
Epoch   6 Batch  287/538 - Train Accuracy: 0.9795, Validation Accuracy: 0.9810, Loss: 0.0139
Epoch   6 Batch  288/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9766, Loss: 0.0163
Epoch   6 Batch  289/538 - Train Accuracy: 0.9745, Validation Accuracy

Epoch   6 Batch  368/538 - Train Accuracy: 0.9907, Validation Accuracy: 0.9723, Loss: 0.0157
Epoch   6 Batch  369/538 - Train Accuracy: 0.9793, Validation Accuracy: 0.9734, Loss: 0.0163
Epoch   6 Batch  370/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9664, Loss: 0.0170
Epoch   6 Batch  371/538 - Train Accuracy: 0.9756, Validation Accuracy: 0.9675, Loss: 0.0179
Epoch   6 Batch  372/538 - Train Accuracy: 0.9742, Validation Accuracy: 0.9677, Loss: 0.0174
Epoch   6 Batch  373/538 - Train Accuracy: 0.9736, Validation Accuracy: 0.9652, Loss: 0.0160
Epoch   6 Batch  374/538 - Train Accuracy: 0.9771, Validation Accuracy: 0.9650, Loss: 0.0170
Epoch   6 Batch  375/538 - Train Accuracy: 0.9777, Validation Accuracy: 0.9629, Loss: 0.0156
Epoch   6 Batch  376/538 - Train Accuracy: 0.9828, Validation Accuracy: 0.9616, Loss: 0.0178
Epoch   6 Batch  377/538 - Train Accuracy: 0.9820, Validation Accuracy: 0.9583, Loss: 0.0235
Epoch   6 Batch  378/538 - Train Accuracy: 0.9836, Validation Accuracy

Epoch   6 Batch  458/538 - Train Accuracy: 0.9887, Validation Accuracy: 0.9620, Loss: 0.0171
Epoch   6 Batch  459/538 - Train Accuracy: 0.9764, Validation Accuracy: 0.9618, Loss: 0.0118
Epoch   6 Batch  460/538 - Train Accuracy: 0.9702, Validation Accuracy: 0.9641, Loss: 0.0181
Epoch   6 Batch  461/538 - Train Accuracy: 0.9753, Validation Accuracy: 0.9689, Loss: 0.0196
Epoch   6 Batch  462/538 - Train Accuracy: 0.9775, Validation Accuracy: 0.9711, Loss: 0.0147
Epoch   6 Batch  463/538 - Train Accuracy: 0.9711, Validation Accuracy: 0.9732, Loss: 0.0179
Epoch   6 Batch  464/538 - Train Accuracy: 0.9838, Validation Accuracy: 0.9719, Loss: 0.0148
Epoch   6 Batch  465/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9743, Loss: 0.0170
Epoch   6 Batch  466/538 - Train Accuracy: 0.9680, Validation Accuracy: 0.9751, Loss: 0.0182
Epoch   6 Batch  467/538 - Train Accuracy: 0.9825, Validation Accuracy: 0.9762, Loss: 0.0171
Epoch   6 Batch  468/538 - Train Accuracy: 0.9863, Validation Accuracy

Epoch   7 Batch   11/538 - Train Accuracy: 0.9758, Validation Accuracy: 0.9553, Loss: 0.0159
Epoch   7 Batch   12/538 - Train Accuracy: 0.9713, Validation Accuracy: 0.9576, Loss: 0.0358
Epoch   7 Batch   13/538 - Train Accuracy: 0.9799, Validation Accuracy: 0.9648, Loss: 0.0220
Epoch   7 Batch   14/538 - Train Accuracy: 0.9783, Validation Accuracy: 0.9661, Loss: 0.0145
Epoch   7 Batch   15/538 - Train Accuracy: 0.9678, Validation Accuracy: 0.9611, Loss: 0.0217
Epoch   7 Batch   16/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9611, Loss: 0.0207
Epoch   7 Batch   17/538 - Train Accuracy: 0.9867, Validation Accuracy: 0.9641, Loss: 0.0218
Epoch   7 Batch   18/538 - Train Accuracy: 0.9740, Validation Accuracy: 0.9604, Loss: 0.0261
Epoch   7 Batch   19/538 - Train Accuracy: 0.9848, Validation Accuracy: 0.9606, Loss: 0.0222
Epoch   7 Batch   20/538 - Train Accuracy: 0.9762, Validation Accuracy: 0.9627, Loss: 0.0202
Epoch   7 Batch   21/538 - Train Accuracy: 0.9733, Validation Accuracy

Epoch   7 Batch  100/538 - Train Accuracy: 0.9822, Validation Accuracy: 0.9602, Loss: 0.0119
Epoch   7 Batch  101/538 - Train Accuracy: 0.9848, Validation Accuracy: 0.9666, Loss: 0.0181
Epoch   7 Batch  102/538 - Train Accuracy: 0.9676, Validation Accuracy: 0.9679, Loss: 0.0219
Epoch   7 Batch  103/538 - Train Accuracy: 0.9829, Validation Accuracy: 0.9672, Loss: 0.0156
Epoch   7 Batch  104/538 - Train Accuracy: 0.9887, Validation Accuracy: 0.9695, Loss: 0.0153
Epoch   7 Batch  105/538 - Train Accuracy: 0.9647, Validation Accuracy: 0.9718, Loss: 0.0134
Epoch   7 Batch  106/538 - Train Accuracy: 0.9676, Validation Accuracy: 0.9712, Loss: 0.0145
Epoch   7 Batch  107/538 - Train Accuracy: 0.9820, Validation Accuracy: 0.9680, Loss: 0.0259
Epoch   7 Batch  108/538 - Train Accuracy: 0.9795, Validation Accuracy: 0.9643, Loss: 0.0172
Epoch   7 Batch  109/538 - Train Accuracy: 0.9756, Validation Accuracy: 0.9672, Loss: 0.0165
Epoch   7 Batch  110/538 - Train Accuracy: 0.9785, Validation Accuracy

Epoch   7 Batch  189/538 - Train Accuracy: 0.9836, Validation Accuracy: 0.9707, Loss: 0.0187
Epoch   7 Batch  190/538 - Train Accuracy: 0.9658, Validation Accuracy: 0.9727, Loss: 0.0238
Epoch   7 Batch  191/538 - Train Accuracy: 0.9807, Validation Accuracy: 0.9787, Loss: 0.0127
Epoch   7 Batch  192/538 - Train Accuracy: 0.9877, Validation Accuracy: 0.9762, Loss: 0.0158
Epoch   7 Batch  193/538 - Train Accuracy: 0.9808, Validation Accuracy: 0.9759, Loss: 0.0177
Epoch   7 Batch  194/538 - Train Accuracy: 0.9717, Validation Accuracy: 0.9684, Loss: 0.0190
Epoch   7 Batch  195/538 - Train Accuracy: 0.9868, Validation Accuracy: 0.9668, Loss: 0.0190
Epoch   7 Batch  196/538 - Train Accuracy: 0.9764, Validation Accuracy: 0.9631, Loss: 0.0142
Epoch   7 Batch  197/538 - Train Accuracy: 0.9650, Validation Accuracy: 0.9648, Loss: 0.0182
Epoch   7 Batch  198/538 - Train Accuracy: 0.9717, Validation Accuracy: 0.9654, Loss: 0.0155
Epoch   7 Batch  199/538 - Train Accuracy: 0.9785, Validation Accuracy

Epoch   7 Batch  278/538 - Train Accuracy: 0.9762, Validation Accuracy: 0.9730, Loss: 0.0148
Epoch   7 Batch  279/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9732, Loss: 0.0103
Epoch   7 Batch  280/538 - Train Accuracy: 0.9790, Validation Accuracy: 0.9679, Loss: 0.0139
Epoch   7 Batch  281/538 - Train Accuracy: 0.9768, Validation Accuracy: 0.9703, Loss: 0.0155
Epoch   7 Batch  282/538 - Train Accuracy: 0.9846, Validation Accuracy: 0.9737, Loss: 0.0198
Epoch   7 Batch  283/538 - Train Accuracy: 0.9834, Validation Accuracy: 0.9734, Loss: 0.0217
Epoch   7 Batch  284/538 - Train Accuracy: 0.9893, Validation Accuracy: 0.9714, Loss: 0.0158
Epoch   7 Batch  285/538 - Train Accuracy: 0.9885, Validation Accuracy: 0.9725, Loss: 0.0162
Epoch   7 Batch  286/538 - Train Accuracy: 0.9721, Validation Accuracy: 0.9727, Loss: 0.0240
Epoch   7 Batch  287/538 - Train Accuracy: 0.9797, Validation Accuracy: 0.9744, Loss: 0.0144
Epoch   7 Batch  288/538 - Train Accuracy: 0.9771, Validation Accuracy

Epoch   7 Batch  367/538 - Train Accuracy: 0.9879, Validation Accuracy: 0.9723, Loss: 0.0124
Epoch   7 Batch  368/538 - Train Accuracy: 0.9850, Validation Accuracy: 0.9721, Loss: 0.0115
Epoch   7 Batch  369/538 - Train Accuracy: 0.9693, Validation Accuracy: 0.9716, Loss: 0.0121
Epoch   7 Batch  370/538 - Train Accuracy: 0.9787, Validation Accuracy: 0.9718, Loss: 0.0138
Epoch   7 Batch  371/538 - Train Accuracy: 0.9900, Validation Accuracy: 0.9680, Loss: 0.0140
Epoch   7 Batch  372/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9618, Loss: 0.0161
Epoch   7 Batch  373/538 - Train Accuracy: 0.9783, Validation Accuracy: 0.9590, Loss: 0.0118
Epoch   7 Batch  374/538 - Train Accuracy: 0.9789, Validation Accuracy: 0.9634, Loss: 0.0141
Epoch   7 Batch  375/538 - Train Accuracy: 0.9699, Validation Accuracy: 0.9691, Loss: 0.0177
Epoch   7 Batch  376/538 - Train Accuracy: 0.9824, Validation Accuracy: 0.9721, Loss: 0.0156
Epoch   7 Batch  377/538 - Train Accuracy: 0.9740, Validation Accuracy

Epoch   7 Batch  456/538 - Train Accuracy: 0.9810, Validation Accuracy: 0.9705, Loss: 0.0249
Epoch   7 Batch  457/538 - Train Accuracy: 0.9809, Validation Accuracy: 0.9709, Loss: 0.0142
Epoch   7 Batch  458/538 - Train Accuracy: 0.9825, Validation Accuracy: 0.9661, Loss: 0.0158
Epoch   7 Batch  459/538 - Train Accuracy: 0.9939, Validation Accuracy: 0.9689, Loss: 0.0098
Epoch   7 Batch  460/538 - Train Accuracy: 0.9821, Validation Accuracy: 0.9689, Loss: 0.0144
Epoch   7 Batch  461/538 - Train Accuracy: 0.9825, Validation Accuracy: 0.9714, Loss: 0.0143
Epoch   7 Batch  462/538 - Train Accuracy: 0.9788, Validation Accuracy: 0.9664, Loss: 0.0150
Epoch   7 Batch  463/538 - Train Accuracy: 0.9750, Validation Accuracy: 0.9675, Loss: 0.0175
Epoch   7 Batch  464/538 - Train Accuracy: 0.9873, Validation Accuracy: 0.9688, Loss: 0.0130
Epoch   7 Batch  465/538 - Train Accuracy: 0.9695, Validation Accuracy: 0.9714, Loss: 0.0123
Epoch   7 Batch  466/538 - Train Accuracy: 0.9742, Validation Accuracy

Epoch   8 Batch    9/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9814, Loss: 0.0156
Epoch   8 Batch   10/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9826, Loss: 0.0146
Epoch   8 Batch   11/538 - Train Accuracy: 0.9859, Validation Accuracy: 0.9814, Loss: 0.0146
Epoch   8 Batch   12/538 - Train Accuracy: 0.9822, Validation Accuracy: 0.9806, Loss: 0.0133
Epoch   8 Batch   13/538 - Train Accuracy: 0.9909, Validation Accuracy: 0.9831, Loss: 0.0116
Epoch   8 Batch   14/538 - Train Accuracy: 0.9834, Validation Accuracy: 0.9830, Loss: 0.0119
Epoch   8 Batch   15/538 - Train Accuracy: 0.9860, Validation Accuracy: 0.9821, Loss: 0.0104
Epoch   8 Batch   16/538 - Train Accuracy: 0.9846, Validation Accuracy: 0.9808, Loss: 0.0144
Epoch   8 Batch   17/538 - Train Accuracy: 0.9877, Validation Accuracy: 0.9810, Loss: 0.0139
Epoch   8 Batch   18/538 - Train Accuracy: 0.9859, Validation Accuracy: 0.9824, Loss: 0.0178
Epoch   8 Batch   19/538 - Train Accuracy: 0.9812, Validation Accuracy

Epoch   8 Batch   98/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9741, Loss: 0.0139
Epoch   8 Batch   99/538 - Train Accuracy: 0.9836, Validation Accuracy: 0.9748, Loss: 0.0129
Epoch   8 Batch  100/538 - Train Accuracy: 0.9932, Validation Accuracy: 0.9780, Loss: 0.0113
Epoch   8 Batch  101/538 - Train Accuracy: 0.9811, Validation Accuracy: 0.9787, Loss: 0.0186
Epoch   8 Batch  102/538 - Train Accuracy: 0.9744, Validation Accuracy: 0.9775, Loss: 0.0178
Epoch   8 Batch  103/538 - Train Accuracy: 0.9872, Validation Accuracy: 0.9783, Loss: 0.0155
Epoch   8 Batch  104/538 - Train Accuracy: 0.9844, Validation Accuracy: 0.9764, Loss: 0.0103
Epoch   8 Batch  105/538 - Train Accuracy: 0.9775, Validation Accuracy: 0.9698, Loss: 0.0131
Epoch   8 Batch  106/538 - Train Accuracy: 0.9754, Validation Accuracy: 0.9702, Loss: 0.0117
Epoch   8 Batch  107/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9746, Loss: 0.0156
Epoch   8 Batch  108/538 - Train Accuracy: 0.9898, Validation Accuracy

Epoch   8 Batch  187/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9730, Loss: 0.0140
Epoch   8 Batch  188/538 - Train Accuracy: 0.9896, Validation Accuracy: 0.9755, Loss: 0.0111
Epoch   8 Batch  189/538 - Train Accuracy: 0.9875, Validation Accuracy: 0.9776, Loss: 0.0162
Epoch   8 Batch  190/538 - Train Accuracy: 0.9754, Validation Accuracy: 0.9773, Loss: 0.0170
Epoch   8 Batch  191/538 - Train Accuracy: 0.9829, Validation Accuracy: 0.9741, Loss: 0.0159
Epoch   8 Batch  192/538 - Train Accuracy: 0.9857, Validation Accuracy: 0.9741, Loss: 0.0126
Epoch   8 Batch  193/538 - Train Accuracy: 0.9823, Validation Accuracy: 0.9707, Loss: 0.0163
Epoch   8 Batch  194/538 - Train Accuracy: 0.9770, Validation Accuracy: 0.9682, Loss: 0.0181
Epoch   8 Batch  195/538 - Train Accuracy: 0.9833, Validation Accuracy: 0.9631, Loss: 0.0186
Epoch   8 Batch  196/538 - Train Accuracy: 0.9849, Validation Accuracy: 0.9650, Loss: 0.0136
Epoch   8 Batch  197/538 - Train Accuracy: 0.9760, Validation Accuracy

Epoch   8 Batch  276/538 - Train Accuracy: 0.9832, Validation Accuracy: 0.9652, Loss: 0.0324
Epoch   8 Batch  277/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9613, Loss: 0.0227
Epoch   8 Batch  278/538 - Train Accuracy: 0.9734, Validation Accuracy: 0.9620, Loss: 0.0229
Epoch   8 Batch  279/538 - Train Accuracy: 0.9691, Validation Accuracy: 0.9609, Loss: 0.0221
Epoch   8 Batch  280/538 - Train Accuracy: 0.9775, Validation Accuracy: 0.9593, Loss: 0.0180
Epoch   8 Batch  281/538 - Train Accuracy: 0.9701, Validation Accuracy: 0.9656, Loss: 0.0276
Epoch   8 Batch  282/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9668, Loss: 0.0291
Epoch   8 Batch  283/538 - Train Accuracy: 0.9947, Validation Accuracy: 0.9675, Loss: 0.0306
Epoch   8 Batch  284/538 - Train Accuracy: 0.9787, Validation Accuracy: 0.9654, Loss: 0.0209
Epoch   8 Batch  285/538 - Train Accuracy: 0.9781, Validation Accuracy: 0.9668, Loss: 0.0213
Epoch   8 Batch  286/538 - Train Accuracy: 0.9656, Validation Accuracy

Epoch   8 Batch  365/538 - Train Accuracy: 0.9760, Validation Accuracy: 0.9741, Loss: 0.0136
Epoch   8 Batch  366/538 - Train Accuracy: 0.9865, Validation Accuracy: 0.9775, Loss: 0.0148
Epoch   8 Batch  367/538 - Train Accuracy: 0.9924, Validation Accuracy: 0.9787, Loss: 0.0110
Epoch   8 Batch  368/538 - Train Accuracy: 0.9910, Validation Accuracy: 0.9785, Loss: 0.0111
Epoch   8 Batch  369/538 - Train Accuracy: 0.9859, Validation Accuracy: 0.9787, Loss: 0.0108
Epoch   8 Batch  370/538 - Train Accuracy: 0.9893, Validation Accuracy: 0.9714, Loss: 0.0143
Epoch   8 Batch  371/538 - Train Accuracy: 0.9939, Validation Accuracy: 0.9714, Loss: 0.0129
Epoch   8 Batch  372/538 - Train Accuracy: 0.9828, Validation Accuracy: 0.9712, Loss: 0.0112
Epoch   8 Batch  373/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9688, Loss: 0.0105
Epoch   8 Batch  374/538 - Train Accuracy: 0.9799, Validation Accuracy: 0.9675, Loss: 0.0130
Epoch   8 Batch  375/538 - Train Accuracy: 0.9723, Validation Accuracy

Epoch   8 Batch  454/538 - Train Accuracy: 0.9864, Validation Accuracy: 0.9721, Loss: 0.0130
Epoch   8 Batch  455/538 - Train Accuracy: 0.9853, Validation Accuracy: 0.9711, Loss: 0.0132
Epoch   8 Batch  456/538 - Train Accuracy: 0.9855, Validation Accuracy: 0.9750, Loss: 0.0236
Epoch   8 Batch  457/538 - Train Accuracy: 0.9766, Validation Accuracy: 0.9760, Loss: 0.0130
Epoch   8 Batch  458/538 - Train Accuracy: 0.9805, Validation Accuracy: 0.9735, Loss: 0.0135
Epoch   8 Batch  459/538 - Train Accuracy: 0.9947, Validation Accuracy: 0.9735, Loss: 0.0087
Epoch   8 Batch  460/538 - Train Accuracy: 0.9842, Validation Accuracy: 0.9751, Loss: 0.0134
Epoch   8 Batch  461/538 - Train Accuracy: 0.9821, Validation Accuracy: 0.9787, Loss: 0.0131
Epoch   8 Batch  462/538 - Train Accuracy: 0.9792, Validation Accuracy: 0.9716, Loss: 0.0122
Epoch   8 Batch  463/538 - Train Accuracy: 0.9703, Validation Accuracy: 0.9796, Loss: 0.0158
Epoch   8 Batch  464/538 - Train Accuracy: 0.9955, Validation Accuracy

Epoch   9 Batch    7/538 - Train Accuracy: 0.9879, Validation Accuracy: 0.9760, Loss: 0.0171
Epoch   9 Batch    8/538 - Train Accuracy: 0.9852, Validation Accuracy: 0.9734, Loss: 0.0076
Epoch   9 Batch    9/538 - Train Accuracy: 0.9781, Validation Accuracy: 0.9766, Loss: 0.0119
Epoch   9 Batch   10/538 - Train Accuracy: 0.9828, Validation Accuracy: 0.9789, Loss: 0.0102
Epoch   9 Batch   11/538 - Train Accuracy: 0.9877, Validation Accuracy: 0.9815, Loss: 0.0150
Epoch   9 Batch   12/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9762, Loss: 0.0128
Epoch   9 Batch   13/538 - Train Accuracy: 0.9829, Validation Accuracy: 0.9762, Loss: 0.0121
Epoch   9 Batch   14/538 - Train Accuracy: 0.9930, Validation Accuracy: 0.9762, Loss: 0.0098
Epoch   9 Batch   15/538 - Train Accuracy: 0.9773, Validation Accuracy: 0.9780, Loss: 0.0108
Epoch   9 Batch   16/538 - Train Accuracy: 0.9907, Validation Accuracy: 0.9799, Loss: 0.0139
Epoch   9 Batch   17/538 - Train Accuracy: 0.9889, Validation Accuracy

Epoch   9 Batch   97/538 - Train Accuracy: 0.9768, Validation Accuracy: 0.9712, Loss: 0.0105
Epoch   9 Batch   98/538 - Train Accuracy: 0.9758, Validation Accuracy: 0.9711, Loss: 0.0115
Epoch   9 Batch   99/538 - Train Accuracy: 0.9877, Validation Accuracy: 0.9705, Loss: 0.0105
Epoch   9 Batch  100/538 - Train Accuracy: 0.9889, Validation Accuracy: 0.9719, Loss: 0.0087
Epoch   9 Batch  101/538 - Train Accuracy: 0.9781, Validation Accuracy: 0.9744, Loss: 0.0187
Epoch   9 Batch  102/538 - Train Accuracy: 0.9812, Validation Accuracy: 0.9732, Loss: 0.0178
Epoch   9 Batch  103/538 - Train Accuracy: 0.9831, Validation Accuracy: 0.9771, Loss: 0.0109
Epoch   9 Batch  104/538 - Train Accuracy: 0.9872, Validation Accuracy: 0.9775, Loss: 0.0090
Epoch   9 Batch  105/538 - Train Accuracy: 0.9764, Validation Accuracy: 0.9751, Loss: 0.0113
Epoch   9 Batch  106/538 - Train Accuracy: 0.9869, Validation Accuracy: 0.9776, Loss: 0.0127
Epoch   9 Batch  107/538 - Train Accuracy: 0.9838, Validation Accuracy

Epoch   9 Batch  186/538 - Train Accuracy: 0.9846, Validation Accuracy: 0.9648, Loss: 0.0129
Epoch   9 Batch  187/538 - Train Accuracy: 0.9794, Validation Accuracy: 0.9597, Loss: 0.0113
Epoch   9 Batch  188/538 - Train Accuracy: 0.9822, Validation Accuracy: 0.9629, Loss: 0.0104
Epoch   9 Batch  189/538 - Train Accuracy: 0.9801, Validation Accuracy: 0.9689, Loss: 0.0139
Epoch   9 Batch  190/538 - Train Accuracy: 0.9829, Validation Accuracy: 0.9643, Loss: 0.0161
Epoch   9 Batch  191/538 - Train Accuracy: 0.9838, Validation Accuracy: 0.9668, Loss: 0.0121
Epoch   9 Batch  192/538 - Train Accuracy: 0.9859, Validation Accuracy: 0.9673, Loss: 0.0123
Epoch   9 Batch  193/538 - Train Accuracy: 0.9896, Validation Accuracy: 0.9648, Loss: 0.0140
Epoch   9 Batch  194/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9641, Loss: 0.0155
Epoch   9 Batch  195/538 - Train Accuracy: 0.9892, Validation Accuracy: 0.9625, Loss: 0.0169
Epoch   9 Batch  196/538 - Train Accuracy: 0.9812, Validation Accuracy

Epoch   9 Batch  276/538 - Train Accuracy: 0.9811, Validation Accuracy: 0.9672, Loss: 0.0167
Epoch   9 Batch  277/538 - Train Accuracy: 0.9922, Validation Accuracy: 0.9661, Loss: 0.0083
Epoch   9 Batch  278/538 - Train Accuracy: 0.9811, Validation Accuracy: 0.9723, Loss: 0.0116
Epoch   9 Batch  279/538 - Train Accuracy: 0.9867, Validation Accuracy: 0.9759, Loss: 0.0111
Epoch   9 Batch  280/538 - Train Accuracy: 0.9877, Validation Accuracy: 0.9727, Loss: 0.0099
Epoch   9 Batch  281/538 - Train Accuracy: 0.9863, Validation Accuracy: 0.9753, Loss: 0.0119
Epoch   9 Batch  282/538 - Train Accuracy: 0.9827, Validation Accuracy: 0.9753, Loss: 0.0133
Epoch   9 Batch  283/538 - Train Accuracy: 0.9908, Validation Accuracy: 0.9783, Loss: 0.0114
Epoch   9 Batch  284/538 - Train Accuracy: 0.9828, Validation Accuracy: 0.9759, Loss: 0.0107
Epoch   9 Batch  285/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9783, Loss: 0.0122
Epoch   9 Batch  286/538 - Train Accuracy: 0.9764, Validation Accuracy

Epoch   9 Batch  365/538 - Train Accuracy: 0.9769, Validation Accuracy: 0.9783, Loss: 0.0143
Epoch   9 Batch  366/538 - Train Accuracy: 0.9883, Validation Accuracy: 0.9808, Loss: 0.0130
Epoch   9 Batch  367/538 - Train Accuracy: 0.9908, Validation Accuracy: 0.9799, Loss: 0.0092
Epoch   9 Batch  368/538 - Train Accuracy: 0.9966, Validation Accuracy: 0.9737, Loss: 0.0092
Epoch   9 Batch  369/538 - Train Accuracy: 0.9789, Validation Accuracy: 0.9727, Loss: 0.0122
Epoch   9 Batch  370/538 - Train Accuracy: 0.9844, Validation Accuracy: 0.9739, Loss: 0.0116
Epoch   9 Batch  371/538 - Train Accuracy: 0.9898, Validation Accuracy: 0.9735, Loss: 0.0116
Epoch   9 Batch  372/538 - Train Accuracy: 0.9834, Validation Accuracy: 0.9686, Loss: 0.0124
Epoch   9 Batch  373/538 - Train Accuracy: 0.9768, Validation Accuracy: 0.9631, Loss: 0.0111
Epoch   9 Batch  374/538 - Train Accuracy: 0.9879, Validation Accuracy: 0.9627, Loss: 0.0123
Epoch   9 Batch  375/538 - Train Accuracy: 0.9825, Validation Accuracy

Epoch   9 Batch  454/538 - Train Accuracy: 0.9879, Validation Accuracy: 0.9703, Loss: 0.0151
Epoch   9 Batch  455/538 - Train Accuracy: 0.9840, Validation Accuracy: 0.9718, Loss: 0.0122
Epoch   9 Batch  456/538 - Train Accuracy: 0.9818, Validation Accuracy: 0.9716, Loss: 0.0199
Epoch   9 Batch  457/538 - Train Accuracy: 0.9887, Validation Accuracy: 0.9743, Loss: 0.0123
Epoch   9 Batch  458/538 - Train Accuracy: 0.9920, Validation Accuracy: 0.9645, Loss: 0.0142
Epoch   9 Batch  459/538 - Train Accuracy: 0.9865, Validation Accuracy: 0.9540, Loss: 0.0079
Epoch   9 Batch  460/538 - Train Accuracy: 0.9801, Validation Accuracy: 0.9558, Loss: 0.0180
Epoch   9 Batch  461/538 - Train Accuracy: 0.9803, Validation Accuracy: 0.9670, Loss: 0.0202
Epoch   9 Batch  462/538 - Train Accuracy: 0.9829, Validation Accuracy: 0.9679, Loss: 0.0125
Epoch   9 Batch  463/538 - Train Accuracy: 0.9746, Validation Accuracy: 0.9693, Loss: 0.0181
Epoch   9 Batch  464/538 - Train Accuracy: 0.9900, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [30]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [31]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [32]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    vocab_list = sentence.lower().split()
    vocab_keys = vocab_to_int.keys()
    
    f = lambda vocab: vocab_to_int[vocab] if vocab in vocab_keys else vocab_to_int['<UNK>']
    seq = [f(vocab) for vocab in vocab_list]
    
    return seq

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [43]:
#translate_sentence = 'he saw a old yellow truck .'
translate_sentence = 'he saw a new red car .'

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [6, 68, 118, 31, 202, 15, 61]
  English Words: ['he', 'saw', 'a', 'new', 'red', 'car', '.']

Prediction
  Word Ids:      [176, 322, 266, 69, 308, 32, 11, 299, 1]
  French Words: il a vu une nouvelle voiture rouge . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.